src/components/OpenWith/OpenWith.js

```diff
@@ -86,16 +102,96 @@ class OpenWith extends PureComponent<Props> {
 
+    /**
+     * Fetch app intgrations info needed to render.
+     *
+     * @param {OpenWithIntegrations} openWithintegrations - The available Open With integrations
+     * @return {void}
+     */
+    fetchAppIntegrations = (openWithintegrations: OpenWithIntegrations) => {
+        const { items } = openWithintegrations;
+        const appIntegrationInfoPromises = [];
+
+        items.forEach((integrationItem: OpenWithIntegrationItem) => {
+            const { app_integration: { id } } = integrationItem;
+            const promise = this.fetchAppIntegrationPromise(id);
+            appIntegrationInfoPromises.push(promise);
+        });
+
+        // Update state after we have information for all integrations, which are required to render.
+        Promise.all(appIntegrationInfoPromises)
+            .then((integrations) => {
+                const openWithintegrationsWithIntegrationData = this.completeOpenWithIntegrationData(
+                    openWithintegrations,
+                    integrations
+                );
+
+                this.setState({ integrationsData: openWithintegrationsWithIntegrationData });
+            })
+            .catch(this.fetchErrorCallback);
+    };
+
+    /**
+     * Creates a promise that resolves with the needed app integration fields
+     *
+     * @param {number} id - An app integration ID
+     * @return {Promise} a promise that resolves with app integration data
+     */
+    fetchAppIntegrationPromise = (id: number) => {
+        const { getAppIntegration }: Props = this.props;
+        return new Promise((resolve, reject) => {
+            // Use a passed in getter if provided, replace when no longer needed
+            const appIntegrationGetFunction: Function = getAppIntegration || this.api.getAppIntegrationsAPI(false).get;
+            appIntegrationGetFunction({
+                id,
+                successCallback: (appIntegration) => resolve(appIntegration),
+                errorCallback: (error) => reject(error)
+            });
+        });
+    };
+
+    /**
+     * Handles a fetch error for the open_with_integrations and app_integrations endpoints
+     *
+     * @param {Error} error - An axios fetch error
+     * @return {void}
+     */
+    fetchErrorCallback = (error: Error) => {
+        console.error(error); // eslint-disable-line no-console
+    };
+
+    /**
+     * Completes the app integration mini objects in Open With data with the required fields to render.
+     *
+     * @param {OpenWithIntegrations} openWithintegrations - The available Open With integrations
+     * @param {Array<AppIntegrationItem>} appIntegrations - An array of full app integration items
+     * @return {void}
+     */
+    completeOpenWithIntegrationData = (
+        openWithintegrations: OpenWithIntegrations,
+        appIntegrations: Array<AppIntegrationItem>
+    ): any =>
+        openWithintegrations.items.map((item) => ({
+            ...item,
+            app_integration: appIntegrations.find(
```

Comment:
Should/can `appIntegrations` be an object instead of an array if it's keyed by ID?